Erc 7540 rid updates (#4)
ERC4626-Alliance/ERCs: ERCS/erc-7540.md

| **State**   | **User**                         | **Vault** |
|-------------|---------------------------------|-----------|

> | Pending     | `requestDeposit(assets, receiver, owner, data)` | `asset.transferFrom(msg.sender, vault, assets)`; `pendingDepositRequest[owner] += assets` |

Reviewer: ```suggestion
| Pending     | `requestDeposit(assets, receiver, owner, data)` | `asset.transferFrom(msg.sender, vault, assets)`; `pendingDepositRequest[receiver] += assets` |
```

Reviewer: This is a bit confusing because here it's the `receiver` param of `requestDeposit`, but in the next line it should be the `owner` that gets an increase in claimable balance 🤔

> | Pending     | `requestDeposit(assets, receiver, owner, data)` | `asset.transferFrom(msg.sender, vault, assets)`; `pendingDepositRequest[receiver] += assets` |

Reviewer: Should have caught this before, my bad
```suggestion
| Pending     | `requestDeposit(assets, receiver, owner, data)` | `asset.transferFrom(owner, vault, assets)`; `pendingDepositRequest[receiver] += assets` |
```

> | Claimable   |                                 | *Internal Request fulfillment*:  `pendingDepositRequest[msg.sender] -= assets`; `claimableDepositRequest[owner] += assets` |

Reviewer: ```suggestion
| Claimable   |                                 | *Internal Request fulfillment*:  `pendingDepositRequest[owner] -= assets`; `claimableDepositRequest[owner] += assets` |
```

> | Claimed     | `deposit(assets, receiver)`      | `claimableDepositRequest[msg.sender] -= assets`; `vault.balanceOf[receiver] += shares` |

Reviewer: ```suggestion
| Claimed     | `deposit(assets, receiver)`      | `claimableDepositRequest[owner] -= assets`; `vault.balanceOf[receiver] += shares` |
```

### Request Ids
The request ID (`rid`) of a request is returned by the corresponding `requestDeposit` and `requestRedeem` functions.

Multiple requests may have the same `rid`, so a given Request is discriminated by both the `rid` and the `owner`. 

Requests of the same `rid` MUST be fungible with each other (except in the special case `rid == 0` described below). I.e. all Requests with the same `rid` MUST transition from Pending to Claimable at the same time and receive the same exchange rate between `assets` and `shares`.

If a Request becomes partially claimable, all requests of the same `rid` MUST become claimable at the same pro rata rate.

There are no assumptions or requirements of requests with different `rid`. I.e. they MAY transition to Claimable at different times and exchange rates with no ordering or correlation enforced in any way.

When `rid==0`, the Vault MUST use purely the `owner` to discriminate the request state. The Pending and Claimable state of multiple requests from the same `owner` would be aggregated. If a Vault returns `0` for the `rid` of any request, it MUST return `0` for all requests.

### Methods

#### requestDeposit

> Transfers `assets` from `msg.sender` into the Vault and submits a Request for asynchronous `deposit/mint`. This places the Request in Pending state, with a corresponding increase in `pendingDepositRequest` for the amount `assets`. 

Reviewer 2: ```suggestion
Transfers `assets` from `msg.sender` into the Vault and submits a Request for asynchronous `deposit`. This places the Request in Pending state, with a corresponding increase in `pendingDepositRequest` for the amount `assets`. 
```

Should we remove the word mint as really we expect the deposit function to be called when claimable to avoid confusion.

> The output `requestId` is used to partially discriminate the request along with the `receiver`. See "Request Ids" section for more info.

Author: ```suggestion
The output `requestId` is used to partially discriminate the request along with the `receiver`. See [Request Ids](#request-ids) section for more info.
```

> The output `requestId` is used to partially descriminate the request along with the `receiver`. See "Request Ids" section for more info.

Reviewer: ```suggestion
The output `requestId` is used to partially discriminate the request along with the `receiver`. See "Request Ids" section for more info.
```

> The output `requestId` is used to partially descriminate the request along with the `receiver`. See "Request Ids" section for more info.

Reviewer: ```suggestion
The output `requestId` is used to partially descriminate the request along with the `receiver`. See [Request Ids](#request-ids) section for more info.
```

> If the length of `data` is not 0, the Request MUST send an `onERC7540DepositReceived` callback to `receiver` following the interface of ERC7540DepositReceiver described in "Request Callbacks" section. If the length of `data` is 0, the Request MUST NOT send a callback.

Reviewer: ```suggestion
If the length of `data` is not 0, the Request MUST send an `onERC7540DepositReceived` callback to `receiver` following the interface of ERC7540DepositReceiver described in [Request Callbacks](#request-callbacks) section. If the length of `data` is 0, the Request MUST NOT send a callback.
```

  inputs:
    - name: assets
      type: uint256
    - name: receiver
      type: address
    - name: owner
      type: address
    - name: data
      type: bytes
  outputs:

>     - name: rid

Reviewer: nit: I would probably write out full `requestId` names in the formal spec, rather than the acronym `rid`

#### requestRedeem

Assumes control of `shares` from `owner` and submits a Request for asynchronous `redeem`. This places the Request in Pending state, with a corresponding increase in `pendingRedeemRequest` for the amount `shares`. 

> The output `requestId` is used to partially descriminate the request along with the `receiver`. See "Request Ids" section for more info.

Reviewer: ```suggestion
The output `requestId` is used to partially discriminate the request along with the `receiver`. See [Request Ids](#request-ids) section for more info.
```

MAY support either a locking or a burning mechanism for `shares` depending on the Vault implemention. 

If a Vault uses a locking mechanism for `shares`, those `shares` MUST be burned from the Vault balance before or upon claiming the Request.

MUST support a redeem Request flow where the control of `shares` is taken from `owner` directly where `msg.sender` has ERC-20 approval over the `shares` of `owner`.

> If the length of `data` is not 0, the Request MUST send an `onERC7540RedeemReceived` callback to `receiver` following the interface of ERC7540RedeemReceiver described in "Request Callbacks" section. If the length of `data` is 0, the Request MUST NOT send a callback.

Reviewer: ```suggestion
If the length of `data` is not 0, the Request MUST send an `onERC7540RedeemReceived` callback to `receiver` following the interface of ERC7540RedeemReceiver described in [Request Callbacks](#request-callbacks) section. If the length of `data` is 0, the Request MUST NOT send a callback.
```

> Assumes control of `shares` from `owner` and submits a Request for asynchronous `redeem/withdraw`. This places the Request in Pending state, with a corresponding increase in `pendingRedeemRequest` for the amount `shares`. 

Reviewer 2: Similar to requestDeposit 
```suggestion
Assumes control of `shares` from `owner` and submits a Request for asynchronous `redeem`. This places the Request in Pending state, with a corresponding increase in `pendingRedeemRequest` for the amount `shares`. 
```

should we remove the word withdraw as we expect people to call the redeem method?

    - name: sender
      indexed: true
      type: address
    - name: receiver
      indexed: true
      type: address

>     - name: rid

Reviewer: This should include `owner`, right? Since it is not necessarily equivalent to `sender`
```suggestion
      type: address
    - name: owner
      indexed: true
      type: address
    - name: rid
```

Author: yes, I updated this section

>       indexed: true

Reviewer: Maybe makes more sense to put this after `sender` and before `receiver`, to match the order in the view methods?

Author: Its tricky because sender is not indexed, we're maxed at 3 indexed entries

Reviewer: True, makes sense!

  type: event

  inputs:
    - name: receiver
      indexed: true
      type: address
    - name: owner
      indexed: true
      type: address
    - name: rid
      indexed: true
      type: uint256

>     - name: sender

Reviewer: Why is `sender` here towards the end, compared to the `DepositRequest` event?

Author: I fixed depositRequest to use the same order

    - name: rid
      type: uint256
    - name: data
      type: bytes
  outputs:
    - name: interfaceId
      type: bytes4
```

#### ERC7540RedeemReceiver
The interface to be called on `requestRedeem`.

`operator` is the `msg.sender` of the original `requestRedeem` call. `owner` is the `owner` of the `requestRedeem`. `rid` is the output `rid` of the `requestRedeem` and `data` is the `data` of the `requestRedeem`.

This function MUST return `TODO` upon successful execution of the callback.

```yaml
- name: onERC7540RedeemReceived
  type: function

  inputs:
    - name: operator
      type: address
    - name: owner
      type: address
    - name: rid
      type: uint256
    - name: data
      type: bytes
  outputs:
    - name: interfaceId
      type: bytes4
```

### [ERC-165](./eip-165.md) support

Smart contracts implementing this standard MUST implement the [ERC-165](./eip-165.md) `supportsInterface` function.

Asynchronous deposit Vaults MUST return the constant value `true` if TODO `0xea446681` is passed through the `interfaceID` argument.

Asynchronous redemption Vaults MUST return the constant value `true` if TODO `0x2e9dd5bd` is passed through the `interfaceID` argument.

> ERC7540DepositReceiver contracts MUST return the constant value `true` if TODO is passed through the `interfaceID` argument.

Reviewer: Do the receiver contracts need to implement ERC-165? Just ensuring that they return `bytes4(keccak256("onERC7540DepositReceived(address,address,uint256,bytes)"))` on `onERC7540DepositReceived` calls, and vice verse for redeem, is sufficient, I think?

Author: You're right, lets leave it out and I'll reword

  type: event

  inputs:
    - name: receiver
      indexed: true
      type: address
    - name: owner
      indexed: true
      type: address
    - name: requestId
      indexed: true
      type: uint256
    - name: sender
      indexed: false
      type: uint256
    - name: assets
      indexed: false
      type: uint256
```

### Request Callbacks

All methods which initiate a request (including `requestId==0`) include a `data` parameter, which if nonzero length MUST send a callback to the receiver. 

There are two interfaces, `ERC7540DepositReceiver` and `ERC7540RedeemReceiver` which each define the single callback method to be called.

#### ERC7540DepositReceiver
The interface to be called on `requestDeposit`.

`operator` is the `msg.sender` of the original `requestDeposit` call. `owner` is the `owner` of the `requestDeposit`. `requestId` is the output `requestId` of the `requestDeposit` and `data` is the `data` of the `requestDeposit`.

> This function MUST return `TODO` upon successful execution of the callback.

Reviewer: ```suggestion
This function MUST return `0xe74d2a41` upon successful execution of the callback.
```

```yaml
- name: onERC7540DepositReceived
  type: function

  inputs:
    - name: operator
      type: address
    - name: owner
      type: address
    - name: requestId
      type: uint256
    - name: data
      type: bytes
  outputs:
    - name: interfaceId
      type: bytes4
```

#### ERC7540RedeemReceiver
The interface to be called on `requestRedeem`.

`operator` is the `msg.sender` of the original `requestRedeem` call. `owner` is the `owner` of the `requestRedeem`. `requestId` is the output `requestId` of the `requestRedeem` and `data` is the `data` of the `requestRedeem`.

> This function MUST return `TODO` upon successful execution of the callback.

Reviewer: ```suggestion
This function MUST return `0x0102fde4` upon successful execution of the callback.
```

```yaml
- name: onERC7540RedeemReceived
  type: function

  inputs:
    - name: operator
      type: address
    - name: owner
      type: address
    - name: requestId
      type: uint256
    - name: data
      type: bytes
  outputs:
    - name: interfaceId
      type: bytes4
```

### [ERC-165](./eip-165.md) support

Smart contracts implementing this Vault standard MUST implement the [ERC-165](./eip-165.md) `supportsInterface` function.

> Asynchronous deposit Vaults MUST return the constant value `true` if TODO `0xea446681` is passed through the `interfaceID` argument.

Reviewer: ```suggestion
Asynchronous deposit Vaults MUST return the constant value `true` if `0x1683f250` is passed through the `interfaceID` argument.
```

Reviewer: See https://github.com/centrifuge/liquidity-pools/blob/requests/test/LiquidityPool.t.sol#L98

> Asynchronous redemption Vaults MUST return the constant value `true` if TODO `0x2e9dd5bd` is passed through the `interfaceID` argument.

Reviewer: ```suggestion
Asynchronous redemption Vaults MUST return the constant value `true` if `0x0899cb0b` is passed through the `interfaceID` argument.
```

## Rationale

### Including Request IDs but Not Including a Claim by ID method
Requests in an Asynchronous Vault have properties of NFTs or Semi-Fungible tokens due to their asynchronicity. However, trying to pigeonhole all ERC-7540 Vaults into supporting [ERC-721](./eip-721) or [ERC-1155](./erc-1155) for Requests would create too much interface bloat. 

Using both an id and address to discriminate Requests allows for any of these use cases to be developed at an external layer without adding too much complexity to the core interface.

> Certain Vaults especially `requestId==0` cases benefit from using the underlying ERC4626 methods for claiming because there is no discrimination at the `requestId` level. This standard is written primarily with those use cases in mind. A future standard can optimize for nonzero RID with support for claiming and transferring requests discriminated also with an `requestId`.

Reviewer: ```suggestion
Certain Vaults especially `requestId==0` cases benefit from using the underlying [ERC-4626](./eip-4626) methods for claiming because there is no discrimination at the `requestId` level. This standard is written primarily with those use cases in mind. A future standard can optimize for nonzero request ID with support for claiming and transferring requests discriminated also with an `requestId`.
```

## Reference Implementation

```solidity
    // This code snippet is incomplete pseudocode used for example only and is no way intended to be used in production or guaranteed to be secure

    mapping(address => uint256) public pendingDepositRequest;
    
    mapping(address => uint256) public claimableDepositRequest;

    function requestDeposit(uint256 assets, address receiver, address owner, bytes calldata data) external returns (uint256 requestId) {
        require(assets != 0);
        require(owner == msg.sender);

        requestId = 0; // no requestId associated with this request

>         asset.safeTransferFrom(msg.sender, address(this), assets); // asset here is the Vault underlying asset

Reviewer: ```suggestion
        asset.safeTransferFrom(owner, address(this), assets); // asset here is the Vault underlying asset
```

        pendingDepositRequest[owner] += assets;

        // Perform the callback
        if (data.length != 0) {

>           ERC7540Receiver(receiver).onERC7540DepositReceived(msg.sender, owner, requestId, data);

Reviewer: ```suggestion
          require(ERC7540Receiver(receiver).onERC7540DepositReceived(msg.sender, owner, requestId, data) == ERC7540Receiver.onERC7540DepositReceived.selector, "receiver failed");
```

        }

>         emit DepositRequest(msg.sender, operator, requestId, assets);

Reviewer: ```suggestion
        emit DepositRequest(receiver, owner, requestId, msg.sender, assets);
```

        require(assets != 0);
        require(owner == msg.sender);

        rid = 0; // no rid associated with this request

        asset.safeTransferFrom(msg.sender, address(this), assets); // asset here is the Vault underlying asset

        pendingDepositRequest[owner] += assets;

        // Perform the callback
        if (data.length != 0) {
          ERC7540Receiver(receiver).onERC7540DepositReceived(msg.sender, owner, rid, data);
        }

        emit DepositRequest(msg.sender, operator, rid, assets);
        return rid;
    }

    /**
     * Include some arbitrary transition logic here from Pending to Claimable
     */

    function deposit(uint256 assets, address receiver) external returns (uint256 shares) {
        require(assets != 0);

        claimableDepositRequest[msg.sender] -= assets; // underflow would revert if not enough claimable assets

        shares = convertToShares(assets); // this naive example uses the instantaneous exchange rate. It may be more common to use the rate locked in upon Claimable stage.

        balanceOf[receiver] += shares;

        emit Deposit(msg.sender, receiver, assets, shares);
    }

```

> ## Security Considerations

Reviewer 3: Think this section is a bit choppy which could lead to some good info being glossed over. Suggest re-writing as follows:

In general, asynchronicity concerns make state transitions in the Vault much more complex and vulnerable to security risks. Access control on Vault operations, clear documentation of state transitioning, and invariant checks should all be performed to mitigate these risks. For example: 

- The view methods for viewing Pending and Claimable request states (e.g. `pendingDepositRequest`) are estimates useful for display purposes but can be outdated. The inability to know the final exchange rate will be on any Request requires users to trust the implementation of the asynchronous Vault in the computation of the exchange rate and fulfillment of their Request.

- Shares or assets locked for Requests can be stuck in the Pending state. Vaults may elect to allow for fungibility of pending claims or implement some cancellation functionality to protect users. 

Lastly, it is worth highlighting again here that the Claim functions for any asynchronous flows MUST enforce that `msg.sender == owner` to prevent theft of Claimable `assets` or `shares`.
